Feature/statistics storage

xtuchyna opened the pull request:
## Related Issues and Dependencies

…

## This introduces a breaking change

- [ ] Yes
- [X] No

## This PR implements
Visualization option -v is now set to launch a Dash application that can be accessed by going to localhost:805o

Another major change in this PR is, that from now on we analyse also active (opened) issues and pullrequests.

PR also contians refactorings in various modules.

pacospace commented on srcopsmetrics/cli.py:
```diff
@@ -99,13 +102,18 @@ def cli(
+    if visualize_statistics and repository is not None:
+        visualize_project_results(project=repository, is_local=is_local)
+    elif visualize_statistics and organization is not None:
+        # TODO: maybe implemnet specific visualization for organization?
```
what about adding a warning here?

xtuchyna replied:
now raises NotImplementedError

pacospace commented on srcopsmetrics/storage.py:
```diff
@@ -98,37 +141,36 @@ def get_ceph_store(self) -> CephStore:
 
         Arguments:
-            repo_path {Path} -- path of the inspected github repository
+            file_ath {Path} -- path of the inspected github repository
```
`file_path`

xtuchyna replied:
not sure what you meant
is it correct now?

pacospace replied:
I read `file_ath` :D

xtuchyna replied:
oh damn!

pacospace commented on srcopsmetrics/templates/report.html:
```diff
@@ -0,0 +1,58 @@
+<div>
```
Why this file is in the PR?

pacospace commented on srcopsmetrics/processing.py:
```diff
@@ -511,7 +517,8 @@ def process_issue_labels_to_issue_closers(
         :rtype: { <issue_closer> : { <issue_label> : <label_occurence_in_closed_issues> } }
         """
-        closers = {}
+        closers = {}SK
```
I think `SK` should not be here :)

xtuchyna replied:
maaaan!

pacospace commented on srcopsmetrics/report.py:
```diff
@@ -0,0 +1,172 @@
+import dash_core_components as dcc
+import dash_html_components as html
+import plotly.graph_objects as go
+from pbr.packaging import append_text_list
```
what is pbr?

xtuchyna replied:
removed
I use an auto import tool in vscode and it probably added this...

xtuchyna commented:
> I still have the same issue running `pipenv run python3 srcopsmetrics/cli.py --repository pytorch/serve -p -l`
> 
> ```
> INFO:srcopsmetrics.utils:Cleaning processed knowledge at pytorch/serve
> INFO:srcopsmetrics.storage:Loading knowledge locally
> INFO:srcopsmetrics.storage:Found previous knowledge for pytorch/serve with 70 entities of type Issue
> INFO:srcopsmetrics.storage:Loading knowledge locally
> INFO:srcopsmetrics.storage:Found previous knowledge for pytorch/serve with 121 entities of type PullRequest
> INFO:srcopsmetrics.storage:Loading knowledge locally
> INFO:srcopsmetrics.storage:Found previous knowledge for None with 14 entities of type Processed Knowledge
> INFO:srcopsmetrics.storage:Loading knowledge locally
> INFO:srcopsmetrics.storage:Found previous knowledge for None with 10 entities of type Processed Knowledge
> INFO:srcopsmetrics.storage:Loading knowledge locally
> INFO:srcopsmetrics.storage:No previous knowledge of type Processed Knowledge found
> Traceback (most recent call last):
>   File "srcopsmetrics/cli.py", line 148, in <module>
>     cli()
>   File "/home/fmurdaca/.local/share/virtualenvs/SrcOpsMetrics-OHExtRur/lib/python3.6/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/home/fmurdaca/.local/share/virtualenvs/SrcOpsMetrics-OHExtRur/lib/python3.6/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/home/fmurdaca/.local/share/virtualenvs/SrcOpsMetrics-OHExtRur/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/home/fmurdaca/.local/share/virtualenvs/SrcOpsMetrics-OHExtRur/lib/python3.6/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "srcopsmetrics/cli.py", line 130, in cli
>     Processing(issues=issues, pull_requests=prs).regenerate()
>   File "/home/fmurdaca/work/aicoe/SrcOpsMetrics/srcopsmetrics/processing.py", line 54, in regenerate
>     self.process_issues_closed_by_pr_size()
>   File "/home/fmurdaca/work/aicoe/SrcOpsMetrics/srcopsmetrics/storage.py", line 71, in __call__
>     knowledge = wrapper()
>   File "/home/fmurdaca/work/aicoe/SrcOpsMetrics/srcopsmetrics/storage.py", line 57, in wrapper
>     return self.func(*args, **kwargs)
>   File "/home/fmurdaca/work/aicoe/SrcOpsMetrics/srcopsmetrics/processing.py", line 558, in process_issues_closed_by_pr_size
>     for issue_id in (i for i in self.pull_requests[pr_id]["referenced_issues"] if self.issues[i]["closed_at"]):
>   File "/home/fmurdaca/work/aicoe/SrcOpsMetrics/srcopsmetrics/processing.py", line 558, in <genexpr>
>     for issue_id in (i for i in self.pull_requests[pr_id]["referenced_issues"] if self.issues[i]["closed_at"]):
> KeyError: '205'
> ```

You have to create whole new knowledge for pytorch repository

pacospace commented:
```
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/1
INFO:srcopsmetrics.iterator:[ API requests remaining: 4992 ]
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 70/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 1
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/1
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 71/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 2/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 3/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 4/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 5/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 6/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 7/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 8/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 9/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 10/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 11/12
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 12/12
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 72/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 2
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/2
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 2/2
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 73/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 3
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/3
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 2/3
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 3/3
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 74/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 2
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/2
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 2/2
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 75/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 2
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/2
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 2/2
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 76/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 2
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/2
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 2/2
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 77/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 7
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/7
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 2/7
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 3/7
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 4/7
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 5/7
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 6/7
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 7/7
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 78/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 1
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/1
INFO:srcopsmetrics.iterator:Analysing PullRequest no. 79/157
INFO:srcopsmetrics.github_knowledge:  -num of reviews found: 3
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 1/3
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 2/3
INFO:srcopsmetrics.github_knowledge:      -analysing review no. 3/3
INFO:srcopsmetrics.iterator:[ API requests remaining: 4992 ]
```

Is it normal nothing changed in the number of API requests remaining? @xtuchyna

xtuchyna commented:
@pacospace depending which time specifically the analysis was run.
It can be coincidence that you've run it right before limit was reset and the limit was coincidentally the same as the one before.
I did not encounter such issues.

which probably brings out question, if we should add timestamp on the API rate limit logger message?

pacospace commented:
> @pacospace depending which time specifically the analysis was run.
> It can be coincidence that you've run it right before limit was reset and the limit was coincidentally the same as the one before.
> I did not encounter such issues.
> 
> which probably brings out question, if we should add timestamp on the API rate limit logger message?

Actually I found it always the same, until we move to the analysis of another entitiesand then it will remain the same :)

pacospace commented:
> I like the visualizations!! Great work, but I see some issues:
> 
> 1. Issue in the plot Issue Title
>    ![Screenshot from 2020-05-06 09-16-00](https://user-images.githubusercontent.com/27498679/81149025-afb85a80-8f7d-11ea-952a-bf1c8a8965c8.png)
> 2. Plot are not correct, there is some `sort` missing somewhere in terms of time.
> 
> ![Screenshot from 2020-05-06 09-16-06](https://user-images.githubusercontent.com/27498679/81149108-dd9d9f00-8f7d-11ea-8257-c1edaeb7ad33.png)
> 
> 3. Confusions in title and axis label, median or average?
>    ![Screenshot from 2020-05-06 09-44-57](https://user-images.githubusercontent.com/27498679/81149485-95cb4780-8f7e-11ea-9af5-2776e6166a7d.png)
> 4. How do we select these features for top contributors? Where are the reviews for example?
>    ![Screenshot from 2020-05-06 09-49-40](https://user-images.githubusercontent.com/27498679/81149763-2013ab80-8f7f-11ea-9d4a-cdf0104d44ed.png)
> 5. The issue closed by a contributor has a weight per label? Or do we consider all with same weights?

Point 5 is still not solved, otherwise LGTM!

pacospace commented:
> > I like the visualizations!! Great work, but I see some issues:
> > 
> > 1. Issue in the plot Issue Title
> >    ![Screenshot from 2020-05-06 09-16-00](https://user-images.githubusercontent.com/27498679/81149025-afb85a80-8f7d-11ea-952a-bf1c8a8965c8.png)
> > 2. Plot are not correct, there is some `sort` missing somewhere in terms of time.
> > 
> > ![Screenshot from 2020-05-06 09-16-06](https://user-images.githubusercontent.com/27498679/81149108-dd9d9f00-8f7d-11ea-8257-c1edaeb7ad33.png)
> > 
> > 3. Confusions in title and axis label, median or average?
> >    ![Screenshot from 2020-05-06 09-44-57](https://user-images.githubusercontent.com/27498679/81149485-95cb4780-8f7e-11ea-9af5-2776e6166a7d.png)
> > 4. How do we select these features for top contributors? Where are the reviews for example?
> >    ![Screenshot from 2020-05-06 09-49-40](https://user-images.githubusercontent.com/27498679/81149763-2013ab80-8f7f-11ea-9d4a-cdf0104d44ed.png)
> > 5. The issue closed by a contributor has a weight per label? Or do we consider all with same weights?
> 
> Point 3 is still not solved, otherwise LGTM!

Point 2 is missing name and measurement unit of the axis if any

xtuchyna commented:
@pacospace hope everything's alright, fixed some stats that were processed incorrectly, added some missing labels.

tested right now on pytorch/serve with new collected knowledge, seems everything's alright. does it work on your side also?

xtuchyna commented:
@pacospace ping

xtuchyna commented:
cheers!
